internal/pkg/service/buffer/store/watcher/state.go

```diff
@@ -0,0 +1,177 @@
+package watcher
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/benbjohnson/clock"
+	etcd "go.etcd.io/etcd/client/v3"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store/key"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store/schema"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/prefixtree"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/servicectx"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+type State struct {
+	ctx    context.Context
+	wg     *sync.WaitGroup
+	logger log.Logger
+	client *etcd.Client
+
+	lock        *sync.RWMutex
+	receivers   *objectsState[model.ReceiverBase]
+	exports     *objectsState[model.ExportBase]
+	mappings    *objectsState[model.Mapping]
+	openedFiles *objectsState[model.File]
+	tokens      *objectsState[model.Token]
+}
+
+type objectsState[T any] struct {
+	*prefixtree.Tree[T]
+	initDone <-chan struct{}
+}
+
+type dependencies interface {
+	Clock() clock.Clock
+	Logger() log.Logger
+	Process() *servicectx.Process
+	Schema() *schema.Schema
+	EtcdClient() *etcd.Client
+}
+
+func NewState(d dependencies) *State {
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &State{
+		ctx:    ctx,
+		wg:     &sync.WaitGroup{},
+		logger: d.Logger().AddPrefix("[watcher]"),
+		client: d.EtcdClient(),
+		lock:   &sync.RWMutex{},
+	}
+
+	d.Process().OnShutdown(func() {
+		s.logger.Info("received shutdown request")
+		cancel()
+		s.wg.Wait()
+		s.logger.Info("shutdown done")
+	})
+
+	sm := d.Schema()
+	s.receivers = watch(s, sm.Configs().Receivers().PrefixT())
+	s.exports = watch(s, sm.Configs().Exports().PrefixT())
+	s.mappings = watch(s, sm.Configs().Mappings().PrefixT())
+	s.openedFiles = watch(s, sm.Files().Opened().PrefixT())
+	s.tokens = watch(s, sm.Secrets().Tokens().PrefixT())
+
+	// Wait for initial load
+	startTime := d.Clock().Now()
+	<-s.receivers.initDone
+	<-s.exports.initDone
+	<-s.mappings.initDone
+	<-s.openedFiles.initDone
+	<-s.tokens.initDone
+	s.logger.Infof(`initialized | %s`, d.Clock().Since(startTime))
+	return s
+}
+
+func (s *State) Receiver(k key.ReceiverKey) (out model.Receiver, found bool) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
```

Comment:
Takto mozeme nacitat informacie o `Receiver` z memory a nemusime pri kazdom `import` volat `etcd`.
To by malo vyrazne znizit zataz na `etcd` a trocha zrychlit aj jednotlivy import request.

Reply:
`s.lock` vsetk hodnoty pouzivaju rovnaky zamok + ten sa pouzije aj tu.
Teda sa nemoze stat, ze kym o davame dokopy, tak sa nam to pod rukami zmeni.